### PR TITLE
No std

### PR DESCRIPTION
**Pull request by kpob**

### CODE REVIEW

File: core/src/node.rs [hunk 1]
```diff
@@ -1,9 +1,9 @@
 use crate::types::OdraType;
+use crate::{List, Mapping, Sequence, Variable};
+use alloc::{string::String, vec::Vec};
```
Comment (zie1ony) [inline core/src/node.rs]: I'd avoid `alloc::vec![...]` notation. It doesn't look good.
To import both `vec!` and `Vec` you can do it this way:
```rust
#![no_std]

extern crate alloc;
use alloc::{vec, vec::Vec};

fn main() {
    let a: Vec<u32> = vec![1, 2, 3, 4, 5];
}
```
For some reason `use alloc::vec::{Vec, self};` doesn't work.

File: justfile [hunk 2]
```diff
@@ -2,17 +2,20 @@ default:
     just --list
 
 clippy:
-    cargo clippy --all-targets -- -D warnings
+    cargo clippy --all-targets -- -D warnings
```
Comment (zie1ony) [inline justfile]: Unneeded whitespace at the end of line.

File: justfile [hunk 3]
```diff
@@ -2,17 +2,20 @@ default:
...
 clippy:
-    cargo clippy --all-targets -- -D warnings
+    cargo clippy --all-targets -- -D warnings 
+    cd odra-casper/proxy-caller && cargo clippy --target=wasm32-unknown-unknown -- -D warnings -A clippy::single-component-path-imports
     cd examples && cargo clippy --all-targets -- -D warnings -A clippy::assign-op-pattern
     cd modules && cargo clippy --all-targets -- -D warnings
 
 lint: clippy
     cargo fmt
+    cd odra-casper/proxy-caller && cargo fmt
```
Comment (zie1ony) [inline justfile:12]: I don't think it's needed. `proxy-caller` is in the workspace.

File: justfile [hunk 4]
```diff
@@ -2,17 +2,20 @@ default:
...
     cd examples && cargo fmt
     cd modules && cargo fmt
 
 check-lint: clippy
     cargo fmt -- --check
+    cd odra-casper/proxy-caller && cargo fmt -- --check
```
Comment (zie1ony) [inline justfile:18]: Same here.

File: utils/src/lib.rs [hunk 5]
```diff
@@ -81,8 +87,8 @@ pub fn hex_to_slice(src: &[u8], dst: &mut [u8]) {
 
 /// Joins two parts of a key with the [`KEY_DELIMITER`].
 #[inline]
-pub fn create_key(left: &str, right: &str) -> String {
-    format!("{}{}{}", left, KEY_DELIMITER, right)
+pub fn create_key(left: &str, right: &str) -> alloc::string::String {
```
Comment (zie1ony) [inline utils/src/lib.rs]: Import alloc at the top.

File: types/src/event.rs [hunk 6]
```diff
@@ -10,6 +11,7 @@ pub trait OdraEvent {
     fn emit(self);
     /// Returns the event name.
     fn name() -> String;
+    #[cfg(not(target_arch = "wasm32"))]
```
Comment (zie1ony) [inline types/src/event.rs:14]: Why? This should be needed for events schema registration in `call` in Casper.

File: odra-casper/test-env/src/env.rs [hunk 7]
```diff
@@ -3,7 +3,7 @@
 use std::{
     backtrace::{Backtrace, BacktraceStatus},
     cell::RefCell,
-    collections::HashMap,
+    collections::BTreeMap,
```
Comment (zie1ony) [inline odra-casper/test-env/src/env.rs:6]: It's ok, by why?

File: odra-casper/proxy-caller/src/lib.rs [hunk 8]
```diff
@@ -1,4 +1,5 @@
 #![no_std]
+#![cfg_attr(not(test), feature(core_intrinsics))]
```
Comment (zie1ony) [inline odra-casper/proxy-caller/src/lib.rs:2]: Why this? This lib is not used nor intended to be used in any test.

File: Cargo.toml [hunk 9]
```diff
@@ -13,11 +13,10 @@ members = [
     "odra-casper/shared",
     "odra-casper/test-env",
     "odra-casper/types",
-    "odra-casper/proxy-caller",
```
Comment (zie1ony) [inline Cargo.toml:16]: Why?

File: lang/codegen/src/generator/common.rs [hunk 10]
```diff
@@ -112,7 +112,7 @@ pub(crate) mod mock_vm {
         quote! {
             #[cfg(feature = "mock-vm")]
             impl odra::types::BorshSerialize for #struct_ident {
-                fn serialize<W: std::io::Write>(&self, writer: &mut W) -> std::io::Result<()> {
+                fn serialize<W: odra::types::Write>(&self, writer: &mut W) -> odra::types::Result<()> {
```
Comment (zie1ony) [inline lang/codegen/src/generator/common.rs]: I understand this is is reexport of `borsh`, but `odra::types::Write` is bad path. I'd see here `odra::types::mockvm::borsh::Write`.

File: lang/codegen/src/generator/common.rs [hunk 11]
```diff
@@ -240,7 +240,7 @@ pub(crate) mod casper {
             #[cfg(any(feature = "casper", feature = "casper-livenet"))]
             impl odra::casper::casper_types::bytesrepr::FromBytes for #struct_ident {
                 fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), odra::casper::casper_types::bytesrepr::Error> {
-                    let (_, bytes): (String, _) = odra::casper::casper_types::bytesrepr::FromBytes::from_bytes(bytes)?;
+                    let (_, bytes): (alloc::string::String, _) = odra::casper::casper_types::bytesrepr::FromBytes::from_bytes(bytes)?;
```
Comment (zie1ony) [inline lang/codegen/src/generator/common.rs]: Shouldn't we reexport `alloc` as `odra::reexport::alloc`?

File: mock-vm/types/src/lib.rs [hunk 12]
```diff
@@ -9,7 +9,10 @@ mod uints;
 
 pub use address::Address;
 pub use address::CONTRACT_ADDRESS_PREFIX;
-pub use borsh::{BorshDeserialize, BorshSerialize};
+pub use borsh::{
```
Comment (zie1ony) [inline mock-vm/types/src/lib.rs:12]: Already mentioned earlier to change this reexport.

File: modules/src/lib.rs [hunk 13]
```diff
@@ -1,3 +1,7 @@
+#![cfg_attr(not(feature = "std"), no_std)]
```
Comment (zie1ony) [inline modules/src/lib.rs:1]: +1

File: modules/src/wrapped_native.rs [hunk 14]
```diff
@@ -204,7 +206,7 @@ mod tests {
             (deposit_amount + deposit_amount).into()
         );
         // Then events were emitted.
-        assert_events!(token, Transfer, Deposit, Transfer, Deposit);
+        // assert_events!(token, Transfer, Deposit, Transfer, Deposit);
```
Comment (zie1ony) [inline modules/src/wrapped_native.rs]: why?

File: odra-casper/backend/src/contract_env.rs [hunk 15]
```diff
@@ -144,6 +144,6 @@ pub fn verify_signature(
 }
 
 /// Creates a hash of the given input. Uses default hash for given backend.
-pub fn hash<T: AsRef<[u8]>>(input: T) -> Vec<u8> {
+pub fn hash<T: AsRef<[u8]>>(input: T) -> alloc::vec::Vec<u8> {
```
Comment (zie1ony) [inline odra-casper/backend/src/contract_env.rs]: Avoid `alloc::vec::Vec<u8>`, import it instead.

File: odra-casper/backend/src/lib.rs [hunk 16]
```diff
@@ -1,4 +1,8 @@
+#![no_std]
 //! Generate Casper contract and interact with Casper host.
+#![cfg_attr(not(test), feature(core_intrinsics))]
```
Comment (zie1ony) [inline odra-casper/backend/src/lib.rs:3]: When is it used for tests?

File: odra-casper/backend/src/lib.rs [hunk 17]
```diff
@@ -8,3 +12,14 @@ pub use casper_contract::{
     self,
     contract_api::{runtime, storage}
 };
+
+#[cfg(target_arch = "wasm32")]
```
Comment (zie1ony) [inline odra-casper/backend/src/lib.rs:16]: Do we use it in any other mode than `wasm32`?

File: odra-casper/backend/src/utils.rs [hunk 18]
```diff
@@ -115,7 +116,7 @@ pub fn install_contract(
 
     // Create new contract.
     if is_upgradable {
-        let access_uref_key = format!("{}_access_token", package_hash_key);
+        let access_uref_key = format(format_args!("{}_access_token", package_hash_key));
```
Comment (zie1ony) [inline odra-casper/backend/src/utils.rs]: Why?

**Review by zie1ony — CHANGES_REQUESTED** [review]

File: odra-casper/proxy-caller/src/lib.rs (hunk 8 above)
Comment (kpob) [reply odra-casper/proxy-caller/src/lib.rs:2]: Panic handler requires this feature, and the feature is active NOT in test.

File: odra-casper/backend/src/utils.rs (hunk 18 above)
Comment (kpob) [reply odra-casper/backend/src/utils.rs]: Ok, `fomat!` can be imported from `alloc`

File: odra-casper/backend/src/lib.rs (hunk 17 above)
Comment (kpob) [reply odra-casper/backend/src/lib.rs:16]: Yes, we do. When we build binding files with  `cargo run --bin contracts_build`

File: odra-casper/backend/src/lib.rs (hunk 16 above)
Comment (kpob) [reply odra-casper/backend/src/lib.rs:3]: Panic handler requires this feature, and the feature is active NOT in test.

File: odra-casper/backend/src/contract_env.rs (hunk 15 above)
Comment (kpob) [reply odra-casper/backend/src/contract_env.rs]: Done

File: modules/src/wrapped_native.rs (hunk 14 above)
Comment (kpob) [reply modules/src/wrapped_native.rs]: Unintended

File: utils/src/lib.rs (hunk 5 above)
Comment (kpob) [reply utils/src/lib.rs]: Done

File: justfile (hunk 2 above)
Comment (kpob) [reply justfile]: Done

File: core/src/node.rs (hunk 1 above)
Comment (kpob) [reply core/src/node.rs]: Done

File: lang/codegen/src/generator/common.rs (hunk 11 above)
Comment (kpob) [reply lang/codegen/src/generator/common.rs]: Added the `prelude` mod that gives a  stable interface between `std` and `no_std` env, then we write
```rust
use odra::prelude::vec::Vec;
use odra::prelude::string::String;
```

**Review by zie1ony — APPROVED** [review]

**Pull request merged**